### PR TITLE
Add ROIPooling op

### PR DESCRIPTION
**Pull request by XingHongChenIntel**
Main "ROI" background information, please ref https://arxiv.org/pdf/1803.07066.pdf and https://arxiv.org/pdf/1504.08083.pdf 

According to openvino spec(https://docs.openvinotoolkit.org/latest/openvino_docs_ops_detection_ROIPooling_1.html), the ROIPooling have two pooling method. 

- `max `
this is the basic ROIPooling operation method. please reference the below paper and openvino ngraph implement.
  > https://arxiv.org/pdf/1807.11590v1.pdf    Fig.6 : Illustration of RoI Pooling, RoI Align and PrROI Pooling.
  > https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp

- `bilinear`
this means that interpolating only a single value at each bin center (without pooling). please ref below paper and ngraph implement. 
  >   https://openaccess.thecvf.com/content_ICCV_2017/papers/He_Mask_R-CNN_ICCV_2017_paper.pdf   comment 2
  > https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp

### CODE REVIEW

File: plaidml/bridge/openvino/ops/roi_pooling.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  auto roi_width = std::max(x_2 - x_1, 1.0f);
+  auto roi_height = std::max(y_2 - y_1, 1.0f);
+
+  int kernel_h = static_cast<int>(std::floor(roi_height / pooled_h));
+  int kernel_w = static_cast<int>(std::floor(roi_width / pooled_w));
+
+  edsl::Tensor w_tensor, h_tensor;
+  edsl::InterpolationMode interpolation_mode;
+  if (method != "bilinear") {
+    h_tensor = indices_generator(static_cast<int>(x_1), pooled_h, kernel_h, DType::INT32, "indices_h");
```
Comment (YangleiZouIntel) [inline plaidml/bridge/openvino/ops/roi_pooling.cpp]: Can you use edsl::index here?
please see 
https://github.com/plaidml/plaidml/pull/1548/files#diff-8683f58c8e11af7cc0957be604d50b7ca9f0e6ff336c45b49a2c6c9c0292d74eR68

File: plaidml/bridge/openvino/ops/roi_pooling.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    std::vector<edsl::Tensor> ROI_pools;
+    // 2D input tensor of shape [NUM_ROIS, 5] describing box
+    for (auto index = coords_box.begin(); index != coords_box.end(); index += BOX_ELEMENT_SIZE) {
+      // consisting of 5 element tuples: [batch_id, x_1, y_1, x_2, y_2]
+      auto batch_id = *index;
+      std::vector<float> coord(index + 1, index + BOX_ELEMENT_SIZE);
+      // translate ROI coordinates from their input spatial scale to the scale used when pooling
+      for (int i = 1; i < coord.size(); i++) {
+        coord[i] *= spatial_ratio;
```
Comment (YangleiZouIntel) [inline plaidml/bridge/openvino/ops/roi_pooling.cpp]: Can you use edsl::index to build a Tensor for coord and then pass to crop_resized as a Tensor？ You can build two Tensors it is easier to write code.

Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: ok, I'm trying it.

File: plaidml/bridge/openvino/ops/roi_pooling.cpp (hunk 1 above)
Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: Done, replaced indices_generator by edsl::index op.

File: plaidml/bridge/openvino/ops/roi_pooling.cpp (hunk 2 above)
Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: we use coord to calculate roi_width ans roi_height, then use them to get kernel_h and kernel_w. those two variable would be used as pool op's parameter, and they have to be `int` type. so, if we use edsl::index to build Tensor for coord, it's hard to get `int kernel_h`  and `int kernel_w` in `crop_resized`.

Comment (YangleiZouIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: I see, the pool op is accepting int, we can keep it as it is.

Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: ok

File: plaidml/bridge/openvino/ops/roi_pooling.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+}  // namespace
+
+namespace PlaidMLPlugin {
+
+void registerROIPooling() {
+  registerOp("ROIPooling", [](const Context& ctx) {
+    auto* layer = ngraph::as_type<ngraph::opset4::ROIPooling>(ctx.layer);
+    IE_ASSERT(ctx.operands.size() == 2);
+    auto I = ctx.operands.at(0);
+    auto I_shape = I.compute_shape().sizes();
```
Comment (YangleiZouIntel) [inline plaidml/bridge/openvino/ops/roi_pooling.cpp]: Is I_shape used anywhere?

Comment (YangleiZouIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: we should avoid using compute_shape()

Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: sorry, I forget to remove this, this suppose for slice op, but now we use gather, it already useless. I fixed it.

File: plaidml/bridge/openvino/ops/roi_pooling.cpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+      edsl::Tensor pooled_tensor;
+      if (method != "bilinear") {
+        // translate ROI coordinates from their input spatial scale to the scale used when pooling
+        for (int i = 1; i < coord.size(); i++) {
+          coord[i] = std::round(coord[i] * spatial_ratio);
+        }
+        pooled_tensor = crop_pooling(slice_I, coord, pooled_height, pooled_width);
+      } else {
+        // follow ngraph implement, it don't use spatial_ratio in "bilinear" method.
```
Comment (YangleiZouIntel) [inline plaidml/bridge/openvino/ops/roi_pooling.cpp]: `// follow ngraph implementation, which doesn't use spatial_ratio in "bilinear" method.`

File: plaidml/bridge/openvino/ops/roi_pooling.cpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  auto* ngraph_const = ngraph::as_type<ngraph::op::Constant>(layer->get_input_node_ptr(operand_idx));
+  if (ngraph_const) {
+    return ngraph_const->cast_vector<T>();
+  } else {
+    THROW_IE_EXCEPTION << "Dynamic slicing not currently supported by PlaidML plugin; all of begin, end, and stride "
+                          "must be Constants.";
+  }
+}
+
+edsl::Tensor crop_pooling(edsl::Tensor I, std::vector<float>& coord, int64_t pooled_h, int64_t pooled_w) {
```
Comment (YangleiZouIntel) [inline plaidml/bridge/openvino/ops/roi_pooling.cpp]: You are doing a variation of max pooling here, maybe crop_max_pooling is a better name.

File: plaidml/bridge/openvino/ops/roi_pooling.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    for (auto index = coords_box.begin(); index != coords_box.end(); index += BOX_ELEMENT_SIZE) {
+      // consisting of 5 element tuples: [batch_id, x_1, y_1, x_2, y_2]
+      auto batch_id = *index;
+      std::vector<float> coord(index + 1, index + BOX_ELEMENT_SIZE);
+
+      auto batch_indices = edsl::index({edsl::TensorDim(1)}, 0) + batch_id;
+      auto slice_I = edsl::gather(I, batch_indices).axis(0);
+
+      edsl::Tensor pooled_tensor;
+      if (method != "bilinear") {
```
Comment (YangleiZouIntel) [inline plaidml/bridge/openvino/ops/roi_pooling.cpp]: I think it is better to use switch here, and throws an error if the method is neither bilinear or max. They may add more methods like avg or something else in the future.

File: plaidml/bridge/openvino/ops/roi_pooling.cpp (hunk 4 above)
Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: ok

File: plaidml/bridge/openvino/ops/roi_pooling.cpp (hunk 5 above)
Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: ok. thanks!

File: plaidml/bridge/openvino/ops/roi_pooling.cpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    for (auto index = coords_box.begin(); index != coords_box.end(); index += BOX_ELEMENT_SIZE) {
+      // consisting of 5 element tuples: [batch_id, x_1, y_1, x_2, y_2]
+      auto batch_id = *index;
+      std::vector<float> coord(index + 1, index + BOX_ELEMENT_SIZE);
+
+      auto batch_indices = edsl::index({edsl::TensorDim(1)}, 0) + batch_id;
+      auto slice_I = edsl::gather(I, batch_indices).axis(0);
+
+      edsl::Tensor pooled_tensor;
+      if (method == "max") {
```
Comment (YangleiZouIntel) [inline plaidml/bridge/openvino/ops/roi_pooling.cpp:132]: This is good. I think it is ready to submit to plaidml.

File: plaidml/bridge/openvino/ops/roi_pooling.cpp (hunk 6 above)
Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: thanks for comment, I try to use switch here, but we only can get method as string (from `layer->get_method()`), it means we have to build enum class or map in this file for using switch. 
According to latest openvino upstream, it create `ROIPoolingType` in `ngraph::helper`, then, use this `ROIPoolingType` in test instance, but still translate this `ROIPoolingType` to string in `setup` step. it means` ngraph::opset4::ROIPooling`  always carry `method` as string, so I think maybe it's better we always use method as string in this file. 
I change this code like  `if (method == "max") {...... } else if (method == "bilinear") {..... }else  {THROW_IE_EXCEPTION}`. I'm not sure if this is acceptable?

File: plaidml/bridge/openvino/ops/roi_pooling.cpp (hunk 7 above)
Comment (XingHongChenIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp:132]: ok

File: plaidml/bridge/openvino/ops/roi_pooling.cpp (hunk 6 above)
Comment (YangleiZouIntel) [reply plaidml/bridge/openvino/ops/roi_pooling.cpp]: Yes, I think it is good and is ready to submit to plaidml.

**Pull request closed without merging**